types/seat/wlr_seat.c

```diff
@@ -175,10 +175,18 @@ void wlr_seat_destroy(struct wlr_seat *seat) {
 
 	struct wlr_seat_client *client, *tmp;
 	wl_list_for_each_safe(client, tmp, &seat->clients, link) {
-		struct wl_resource *resource, *next_resource;
-		wl_resource_for_each_safe(resource, next_resource, &client->wl_resources) {
+		struct wl_resource *resource, *next;
+		/* wl_resource_for_each_safe isn't safe to use here, because the last
+		 * wl_resource_destroy will also destroy the head we cannot do the last
+		 * 'next' update that usually is harmless here.
+		 * Work around this by breaking one step ahead
+		 */
+		wl_resource_for_each_safe(resource, next, &client->wl_resources) {
 			// will destroy other resources as well
 			wl_resource_destroy(resource);
+			if (wl_resource_get_link(next) == &client->wl_resources) {
```

Comment:
I think it's too late to do this here, `client` has already been free'd at that time.

Reply:
This is just a play on pointers so won't access memory (`wl_resource_get_link` is basically `return &resource->link`; and just lookign at `&<invalid>->field` is perfectly safe as it'll only return where the address of field would be)

Reply:
Here's the macro defining `wl_resource_for_each_safe`:

```c
#define wl_resource_for_each_safe(resource, tmp, list)					\
	for (resource = 0, tmp = 0,							\
	     resource = wl_resource_from_link((list)->next),	\
	     tmp = wl_resource_from_link((list)->next->next);	\
	     wl_resource_get_link(resource) != (list);				\
	     resource = tmp,							\
	     tmp = wl_resource_from_link(wl_resource_get_link(resource)->next))
```

The condition that is supposed to break the loop is `wl_resource_get_link(resource) != (list)`. I wonder why it causes an invalid read, while your condition doesn't, since they both look very similar?

Reply:
The condition doesn't cause an invalid read, but the "update" part of the loop runs first.

just before bailing out, `wl_resource_for_each_safe` does:
 - `resource = tmp`
 - `tmp = wl_resource_from_link(wl_resource_get_link(resource)->next`
 - `wl_resource_get_link(resource) != (list)`

the tmp assignment that is not used in the check is what causes the invalid memory accesss; my version just moves the check a bit earlier

Reply:
Oh you're right, thanks for the explanation, seems like I need to learn how to loop again